File: k-random-bean-validation/src/test/java/io/github/krandom/validation/SizeAnnotationHandlerTest.java
package io.github.krandom.validation;

import static org.assertj.core.api.Assertions.assertThat;

import io.github.krandom.KRandom;
import jakarta.validation.ConstraintViolation;
import jakarta.validation.Validation;
import jakarta.validation.Validator;
import jakarta.validation.ValidatorFactory;
import jakarta.validation.constraints.Size;
import java.util.List;
import java.util.Map;
import java.util.Set;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class SizeAnnotationHandlerTest {
  private KRandom kRandom;

  @BeforeEach
  void setUp() {
    kRandom = new KRandom();
  }

  @Test
  void generatedBeanShouldBeValidAccordingToValidationConstraints() {
    TestBean testBean = kRandom.nextObject(TestBean.class);

    assertThat(testBean.getTestString()).hasSizeBetween(1, 10);
    assertThat(testBean.getTestList()).hasSizeBetween(1, 10);
    assertThat(testBean.getTestArray()).hasSizeBetween(1, 10);
    assertThat(testBean.getTestMap()).hasSizeBetween(1, 10);
  }

  @Test
  void generatedBeanShouldBeValidUsingBeanValidationApi() {
    TestBean testBean = kRandom.nextObject(TestBean.class);

    Validator validator;
    try (ValidatorFactory validatorFactory = Validation.buildDefaultValidatorFactory()) {
      validator = validatorFactory.getValidator();
    }
    Set<ConstraintViolation<TestBean>> violations = validator.validate(testBean);

    assertThat(violations).isEmpty();
  }

  static class TestBean {
    @Size(min = 1, max = 10)
    private String testString;

    @Size(min = 1, max = 10)
    private List<String> testList;

    @Size(min = 1, max = 10)
    private String[] testArray;

    @Size(min = 1, max = 10)
    private Map<String, String> testMap;

    public Map<String, String> getTestMap() {
      return testMap;
    }

    public String[] getTestArray() {
      return testArray;
    }

    public List<String> getTestList() {
      return testList;
    }

    public String getTestString() {
      return testString;
    }
  }
}

File: bean-validation/src/test/kotlin/io/github/krandom/validation/SizeAnnotationHandlerTest.kt
package io.github.krandom.validation

import io.github.krandom.KRandom
import io.kotest.matchers.collections.shouldBeEmpty
import io.kotest.matchers.ints.shouldBeInRange
import jakarta.validation.Validation
import jakarta.validation.Validator
import jakarta.validation.constraints.Size
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

internal class SizeAnnotationHandlerTest {
  private lateinit var kRandom: KRandom

  @BeforeEach
  fun setUp() {
    kRandom = KRandom()
  }

  @Test
  fun `generated bean should be valid according to validation constraints`() {
    val testBean = kRandom.nextObject(TestBean::class.java)

    testBean.testString.length shouldBeInRange 1..10
    testBean.testList.size shouldBeInRange 1..10
    testBean.testArray.size shouldBeInRange 1..10
    testBean.testMap.size shouldBeInRange 1..10
  }

  @Test
  fun `generated bean should be valid using bean validation api`() {
    val testBean = kRandom.nextObject(TestBean::class.java)

    val validator: Validator
    Validation.buildDefaultValidatorFactory().use { validatorFactory ->
      validator = validatorFactory.validator
    }
    val violations = validator.validate(testBean)

    violations.shouldBeEmpty()
  }

  @Suppress("ArrayInDataClass")
  internal data class TestBean(
    @field:Size(min = 1, max = 10) val testString: String,
    @field:Size(min = 1, max = 10) val testList: List<String>,
    @field:Size(min = 1, max = 10) val testArray: Array<String>,
    @field:Size(min = 1, max = 10) val testMap: Map<String, String>,
  )
}
